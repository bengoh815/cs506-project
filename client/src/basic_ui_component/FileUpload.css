/* FileUpload.css */
#upload {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  text-align: center;
}

#upload input[type="file"] {
  margin: 10px 0;
  cursor: pointer;
}

#upload p {
  text-align: center;
}

#upload ul {
  list-style-type: none;
  padding: 0;
}

#upload ul li {
  margin: 5px 0;
}

#file-input-label {
  font-size: small;
  color: black;
  background-color: lightgray;
  padding: 0.75rem;
  border-radius: 0.3rem;
  cursor: pointer;
  margin-right: 0.5rem;
}
#file-input-label:hover {
  background-color: #f0f0f0;
}

#upload-file-button {
  font-size: small;
  color: black;
  background-color: rgb(144, 238, 144);
  padding: 0.75rem;
  border-radius: 0.3rem;
  cursor: pointer;
  margin-right: 0.5rem;
}
#upload-file-button:hover {
  background-color: rgba(144, 238, 144, 0.5);
}
#upload-file-button:enabled {
  border: black;
}
#upload-file-button:disabled {
  background-color: white;
  border: black;
  color: #a0a0a0;
  cursor: not-allowed;
}
