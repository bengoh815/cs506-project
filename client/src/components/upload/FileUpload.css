/* FileUpload.css */
#upload {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  text-align: center;
}

#upload p {
  text-align: center;
  padding: 0;
}

#upload ul {
  list-style-type: none;
  padding: 0;
}

#upload ul li {
  margin: 5px 0;
}

#choose-file-button {
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
  margin-right: 0.5rem;
}

#upload-file-button {
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
}

.text-input {
  margin-bottom: 1rem;
}

.form-label {
  font-weight: 500;
}
