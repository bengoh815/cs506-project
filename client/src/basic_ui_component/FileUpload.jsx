import React, { useState } from "react";
import "./FileUpload.css";
import { useEffect } from "react";

const FileUpload = () => {
  // File to be uploaded
  const [file, setFile] = useState(null);

  // Whether a file has been selected
  const [fileSelected, setFileSelected] = useState(false);

  // Name of the file to be uploaded
  const [selectedFileName, setSelectedFileName] = useState("No file chosen");

  // Label for the file input
  const [chooseFileLabel, setChooseFileLabel] = useState("Choose a file");

  // Whether a file is being uploaded
  const [isUploading, setIsUploading] = useState(false);

  /**
   * Update the file name when a new file is selected and enables the upload button.
   *
   * @param {*} event - Event object
   */
  const handleFileChange = (event) => {
    let input = event.target.files[0];

    if (input === undefined) {
      setFileSelected(false);
      setChooseFileLabel("Choose a file");
      setSelectedFileName("No file chosen");
      return;
    }

    setFileSelected(true);
    setChooseFileLabel("Change file");
    setFile(input);
    setSelectedFileName(input.name);
  };

  /**
   * Sends file to backend for conversion and updates the UI to reflect the upload progress.
   */
  const handleUpload = () => {
    console.log(file);
    setIsUploading(true);
    setSelectedFileName(`Uploading ${file.name}...`);

    const formData = new FormData();
    formData.append("audio-file", file);

    // TODO: fetch request to backend
    fetch("http://127.0.0.1:5000/api/v1/recordings", {
      method: "POST",
      headers: {
        // TODO: Add headers
      },
      body: formData,
    }).then((response) => {
      console.log("Posting file to backend...");
      console.log(response);

      // Reset the UI after 2 seconds
      setTimeout(() => {
        setIsUploading(false);
        setFileSelected(false);
        setChooseFileLabel("Choose a file");
        setSelectedFileName("No file chosen");
      }, 2000);
    });
  };

  return (
    <div id="upload">
      <h2 id="upload-heading">Upload</h2>

      <p id="file-name">
        <em>{selectedFileName}</em>
      </p>

      {isUploading ? (
        <></>
      ) : (
        <>
          <div>
            <label htmlFor="file-input" id="file-input-label">
              {chooseFileLabel}
            </label>
            <input
              id="file-input"
              type="file"
              onChange={handleFileChange}
              accept=".mp3, .m4a, .wav"
              hidden
            />
            <input
              type="button"
              id="upload-file-button"
              onClick={handleUpload}
              value="Upload"
              disabled={!fileSelected}
            />
          </div>
        </>
      )}

      <p id="file-format-label">Accepted file upload formats:</p>
      <ul>
        <li>.mp3</li>
        <li>.m4a</li>
        <li>.wav</li>
      </ul>
    </div>
  );
};

export default FileUpload;
