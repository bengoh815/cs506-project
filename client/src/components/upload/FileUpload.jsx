/******************************************************************************
 * Filename: FileUpload.jsx
 * Purpose:  A component that allows the user to upload an audio file.
 * Author:   Victor Nguyen & Don Ma
 *
 * Description:
 * This file contains the FileUpload component that allows the user to upload an
 * audio file. The component displays a file input and a button that the user can
 * click to upload the selected file. The component also displays the name of the
 * selected file and the accepted file formats.
 *
 * Usage:
 * To use this component, simply import it into the desired file and render it.
 *
 * Notes:
 * [Any additional notes, considerations, or important information
 * about the file that may be relevant to developers or users.]
 *
 ******************************************************************************/

import React, { useState, useRef } from "react";
import { Button } from "react-bootstrap";
import "./FileUpload.css";
import ConvertFileModal from "./ConvertFileModal";

const FileUpload = () => {
  // File to be uploaded
  const [file, setFile] = useState(null);

  // Whether a file has been selected
  const [fileSelected, setFileSelected] = useState(false);

  // Label for the file input
  const [chooseFileLabel, setChooseFileLabel] = useState("Choose File");

  // Reference to the file input
  const fileInputRef = useRef();

  // State for the file details modal
  const [showFileDetailsModal, setShowFileDetailsModal] = useState(false);

  /**
   * Closes the file details modal.
   */
  const handleClose = () => {
    setShowFileDetailsModal(false);
    setFileSelected(false);
    setChooseFileLabel("Choose File");
    setFile(null);
    fileInputRef.current.value = null;
  };

  /**
   * Opens the file details modal.
   */
  const handleShow = () => {
    setShowFileDetailsModal(true);
  };

  /**
   * Opens the file input dialog when the button is clicked.
   */
  const handleFileInputClick = () => {
    fileInputRef.current.click();
  };

  /**
   * Update the file name when a new file is selected and enables the upload button.
   *
   * @param {*} event - Event object
   */
  const handleFileChange = (event) => {
    let input = event.target.files[0];

    if (input === undefined) {
      setFileSelected(false);
      setChooseFileLabel("Choose File");
      return;
    } else {
      setFileSelected(true);
      setChooseFileLabel("Change File");
      setFile(input);
    }
  };

  // MARK: Helper Functions

  /**
   * Get the abbreviated file name.
   * @param {*} fileName - Name of the file
   * @returns {string} Abbreviated file name
   */
  const getAbbreviatedFileName = (fileName, length) => {
    let fileExtension = fileName.split(".").pop();
    let baseName = fileName.replace(`.${fileExtension}`, "");

    return baseName.length > length
      ? `${baseName.substring(0, length)}... .${fileExtension}`
      : fileName;
  };

  return (
    <div id="upload" data-testid="upload" style={{ padding: "1rem" }}>
      {/* File Details Modal */}
      <ConvertFileModal
        file={file}
        setFile={setFile}
        fileInputRef={fileInputRef}
        handleFileInputClick={handleFileInputClick}
        handleShow={showFileDetailsModal}
        handleClose={handleClose}
        getAbbreviatedFileName={getAbbreviatedFileName}
      />

      {/* Upload & Convert Heading */}
      <h3 id="upload-heading" data-testid="upload-heading">
        Upload & Convert
      </h3>

      <>
        <div>
          {/* File Input - hidden */}
          <input
            ref={fileInputRef}
            type="file"
            onChange={handleFileChange}
            accept=".mp3, .m4a, .wav, .webm"
            hidden
          />

          {/* Choose File Button */}
          <Button
            id="choose-file-button"
            data-testid="choose-file-button"
            onClick={handleFileInputClick}
            variant="secondary"
          >
            {chooseFileLabel}
          </Button>

          {file && fileSelected && (
            /* Upload Button */
            <Button
              id="upload-file-button"
              data-testid="upload-file-button"
              onClick={handleShow}
              variant="primary"
              disabled={!fileSelected}
            >
              Upload {getAbbreviatedFileName(file.name, 10)}
            </Button>
          )}
        </div>
      </>

      {/* Accept File Formats */}
      <p id="file-format-label" data-testid="file-format-label">
        Accepted file upload formats:
      </p>
      <ul
        id="accepted-file-formats-list"
        data-testid="accepted-file-formats-list"
      >
        <li>.mp3</li>
        <li>.webm</li>
        <li>.m4a</li>
        <li>.wav</li>
      </ul>
    </div>
  );
};

export default FileUpload;
