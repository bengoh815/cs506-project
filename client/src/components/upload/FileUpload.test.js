/******************************************************************************
 * Filename: FileUpload.test.js
 * Purpose:  Tests the FileUpload component.
 * Author:   Victor Nguyen
 *
 * Description:
 * This file contains tests for the FileUpload component. The tests ensure that
 * the component and its elements render correctly. A snapshot test is also included
 * to catch unintended changes to the component's structure or behavior.
 *
 * Usage:
 * Run the tests using the command `npm test`.
 *
 * Note:
 * More tests to be added to test actual functionality of the component
 * (e.g. file upload).
 *
 ******************************************************************************/

import { render, screen, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import FileUpload from "./FileUpload";

describe("FileUpload component", () => {
  // Cleans up the DOM after each test to ensure a clean environment.
  afterEach(() => {
    cleanup();
  });

  it("whole file upload component renders", () => {
    // Arrange
    render(<FileUpload />);
    const fileUploadElement = screen.getByTestId("upload");

    // Assert
    expect(fileUploadElement).toBeInTheDocument();
  });

  it("heading renders", () => {
    // Arrange
    render(<FileUpload />);
    const headingElement = screen.getByTestId("upload-heading");

    // Assert
    expect(headingElement).toBeInTheDocument();
  });

  it("correct heading content", () => {
    // Arrange
    render(<FileUpload />);
    const headingElement = screen.getByTestId("upload-heading");

    // Assert
    expect(headingElement.textContent).toBe("Upload & Convert");
  });

  it("choose file button renders", () => {
    // Arrange
    render(<FileUpload />);
    const chooseFileButton = screen.getByTestId("choose-file-button");

    // Assert
    expect(chooseFileButton).toBeInTheDocument();
  });

  it("correct choose file button label when NO file is selected", () => {
    // Arrange
    render(<FileUpload />);
    const chooseFileButton = screen.getByTestId("choose-file-button");

    // Assert
    expect(chooseFileButton).toHaveTextContent("Choose File");
  });

  it("upload file button is hidden on load", () => {
    // Arrange
    render(<FileUpload />);
    const fileUploadButton = screen.queryByTestId("upload-file-button");

    // Assert
    expect(fileUploadButton).not.toBeInTheDocument();
  });

  it("accepted file formats label renders", () => {
    // Arrange
    render(<FileUpload />);
    const acceptedFileFormatsLabel = screen.getByTestId("file-format-label");

    // Assert
    expect(acceptedFileFormatsLabel).toBeInTheDocument();
  });

  it("correct accepted file formats label", () => {
    // Arrange
    render(<FileUpload />);
    const acceptedFileFormatsLabel = screen.getByTestId("file-format-label");

    // Assert
    expect(acceptedFileFormatsLabel).toHaveTextContent(
      "Accepted file upload formats:",
    );
  });

  it("accepted file formats list renders", () => {
    // Arrange
    render(<FileUpload />);
    const acceptedFileFormatsList = screen.getByTestId(
      "accepted-file-formats-list",
    );

    // Assert
    expect(acceptedFileFormatsList).toBeInTheDocument();
  });

  it("correct accepted file formats list content", () => {
    // Arrange
    render(<FileUpload />);
    const acceptedFileFormatsList = screen.getByTestId(
      "accepted-file-formats-list",
    );
    const fileFormats = [".mp3", ".m4a", ".wav", ".webm"];

    // Assert
    fileFormats.forEach((format) => {
      expect(acceptedFileFormatsList).toHaveTextContent(format);
    });
  });
});
