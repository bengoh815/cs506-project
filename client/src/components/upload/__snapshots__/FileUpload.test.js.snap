// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  data-testid="upload"
  id="upload"
  style={
    Object {
      "padding": "1rem",
    }
  }
>
  <h3
    data-testid="upload-heading"
    id="upload-heading"
  >
    Upload & Convert
  </h3>
  <div>
    <input
      accept=".mp3, .m4a, .wav, .webm"
      hidden={true}
      onChange={[Function]}
      type="file"
    />
    <button
      className="btn btn-secondary"
      data-testid="choose-file-button"
      disabled={false}
      id="choose-file-button"
      onClick={[Function]}
      type="button"
    >
      Choose File
    </button>
  </div>
  <p
    data-testid="file-format-label"
    id="file-format-label"
  >
    Accepted file upload formats:
  </p>
  <ul
    data-testid="accepted-file-formats-list"
    id="accepted-file-formats-list"
  >
    <li>
      .mp3
    </li>
    <li>
      .webm
    </li>
    <li>
      .m4a
    </li>
    <li>
      .wav
    </li>
  </ul>
</div>
`;
